packages/regl-worldview/src/commands/GLText.js

```diff
@@ -0,0 +1,397 @@
+      for (const char of charSet) {
+        const { x, y } = charInfo[char];
+        const data = tinySDF.draw(char);
+        for (let i = 0; i < tinySDF.size; i++) {
+          for (let j = 0; j < tinySDF.size; j++) {
+            // if this character is near the right edge, we don't actually copy the whole square of data
+            if (x + j < textureWidth) {
+              textureData[textureWidth * (y + i) + x + j] = data[i * tinySDF.size + j];
+            }
+          }
+        }
+      }
+
+      return { textureData, textureWidth, textureHeight, charInfo };
+    }
+  );
+
+const vert = `
+  precision mediump float;
+
+  uniform mat4 projection, view, billboardRotation;
+  uniform float fontSize;
+  uniform float srcHeight;
+  uniform vec2 atlasSize;
+
+  // per-vertex attributes
+  attribute vec2 texCoord;
+  attribute vec2 position;
+
+  // per-instance (character) attributes
+  attribute vec2 srcOffset;
+  attribute float srcWidth;
+  attribute vec2 destOffset;
+
+  // per-marker attributes
+  attribute vec3 scale;
+  attribute float billboard;
+  attribute vec2 alignmentOffset;
+  attribute float enableOutline;
+  attribute vec4 foregroundColor;
+  attribute vec4 outlineColor;
+  attribute vec3 posePosition;
+  attribute vec4 poseOrientation;
+
+  varying vec2 vTexCoord;
+  varying float vEnableOutline;
+  varying vec4 vForegroundColor;
+  varying vec4 vOutlineColor;
+
+  // rotate a 3d point v by a rotation quaternion q
+  // like applyPose(), but we need to use a custom per-instance pose
+  vec3 rotate(vec3 v, vec4 q) {
+    vec3 temp = cross(q.xyz, v) + q.w * v;
+    return v + (2.0 * cross(q.xyz, temp));
+  }
+
+  void main () {
+    vec2 srcSize = vec2(srcWidth, srcHeight);
+    vec3 markerSpacePos = scale * vec3((destOffset + position * srcSize + alignmentOffset) / fontSize, 0);
+    vec3 pos;
+    if (billboard == 1.0) {
+      pos = (billboardRotation * vec4(markerSpacePos, 1)).xyz + posePosition;
+    } else {
+      pos = rotate(markerSpacePos, poseOrientation) + posePosition;
+    }
+    gl_Position = projection * view * vec4(pos, 1);
+    vTexCoord = (srcOffset + texCoord * srcSize) / atlasSize;
+    vEnableOutline = enableOutline;
+    vForegroundColor = foregroundColor;
+    vOutlineColor = outlineColor;
+  }
+`;
+
+const frag = `
+  #extension GL_OES_standard_derivatives : enable
+  precision mediump float;
+  uniform mat4 projection;
+  uniform sampler2D atlas;
+  uniform float outlineCutoff;
+  uniform float cutoff;
+  varying vec2 vTexCoord;
+  varying float vEnableOutline;
+  varying vec4 vForegroundColor;
+  varying vec4 vOutlineColor;
+  void main() {
+    float dist = texture2D(atlas, vTexCoord).a;
+
+    // fwidth(dist) is used to provide some anti-aliasing. However it's currently only used
+    // between outline and text, not on the outer border, because the alpha blending and
+    // depth test don't work together nicely for partially-transparent pixels.
+    if (vEnableOutline > 0.5) {
```

Comment:
clarifying question -- `vEnableOutline` is only either 1 or 0 right?

Reply:
Yup. When it was a uniform (before I combined all the markers into 1 draw call) I just had it as a boolean. But when I combined them I needed to pass it into the fragment shader with a `varying`, so it started doing interpolation, and some floating point rounding errors meant that `== 1.0` was giving glitchy results.

Reply:
crazy. thanks!